Less invasive versioning \ diffbase \ SyncDAO/VersionNoDAO/StoreAndForwadDAO

mdittmer opened the pull request:
SyncDAO/VersionNoDAO to use generated "Versioned" classes

This change generates "versioned" classes with properties `version_` and `deleted_` for interal use with `SyncDAO` and `VersionNoDAO`. Currently, `VersionNoDAO` expects a versioned `of` (and does not automatically create one, but maybe it should?).

The value of this change is that client code outside of the `SyncDAO` can operate on classes that do not contain versioning information; this becomes an implementation detail hidden inside: (Sync DAO ==> ... VersionNoDAO => ... ==> <storage mechanism>).

marcogbarcellos committed:
ignoring .DS_Store files

marcogbarcellos committed:
fixing find and remove from MDAO

marcogbarcellos committed:
Merge branch 'master' of https://github.com/foam-framework/foam2 into index-find-fix

marcogbarcellos committed:
removing unnecessary variable

marcogbarcellos committed:
removing unnecessary comment

Commit:
Updated javax http to 3.1.0, implemented required methods and HttpContextListener

Commit:
Tracked HttpContextListener.java

marcogbarcellos committed:
Merge branch 'master' of https://github.com/foam-framework/foam2 into index-find-fix

marcogbarcellos committed:
removing get method from altindex

marcogbarcellos committed:
Changing name from treePlan to TreeLookupFindPlan

mdittmer committed:
Test and implement: Partitions/namespaces for DatastoreDAO/JS

All DatastoreDAOs now set "partitionId" [1]. Injecting
DatastoreDAO.__context__.datastoreNamespaceId or DatastoreDAO.namespaceId
causes DAO to use a custom datastore namespace in its "partitionId".

[1] https://cloud.google.com/datastore/docs/reference/rest/v1/PartitionId

marcogbarcellos committed:
adding comparePropertyValue so PropertyInfo is able to compare a specific Type value with a whole object

marcogbarcellos committed:
adding comparePropertyValue method

marcogbarcellos committed:
Merge branch 'master' of https://github.com/foam-framework/foam2 into index-find-fix

marcogbarcellos committed:
change cost to log2(size)

marcogbarcellos committed:
removing whitespaces

Commit:
styling fixes

marcogbarcellos committed:
removing unnecessary method from AbstractsPropertyInfo Type and improving method name from comparePropertyValue to comparePropertyToObject

marcogbarcellos committed:
validating if state is null on creating a TreeLookupFindPlan on TreeIndex

marcogbarcellos committed:
adding java generation of the comparePropertyToObject for PropertyInfo children

marcogbarcellos committed:
changing compare method name

marcogbarcellos committed:
Merge branch 'master' of https://github.com/foam-framework/foam2 into index-find-fix

marcogbarcellos committed:
adding foam header

kgrgreer committed:
Initial check-in of NSpecAware.

kgrgreer committed:
Added NSpecAware support to NanoRouter.

Commit:
Merged upstream and fixed conflict

Commit:
Servlet name is taken from NSpec now

kgrgreer committed:
Merge pull request #556 from nick-prat/tomcat_integration

Tomcat integration

marcogbarcellos committed:
removing unnecesssary cast and validations

marcogbarcellos committed:
Merge branch 'master' of https://github.com/foam-framework/foam2 into index-find-fix

kgrgreer committed:
Merge pull request #542 from marcogbarcellos/index-find-fix

Fixing MDAO.find to look through Id only and adding TreePlan

kgrgreer committed:
Set tableColumns for NSpec.

kgrgreer committed:
Added tableCellFormatters to failed and passed columns Test columns.

kgrgreer committed:
Improved TableView Date formatting.

kgrgreer committed:
Add tableColumns to Menu.

kgrgreer committed:
Added TestBorder with button to run all tests and shows total counts.

kgrgreer committed:
Added missing ;.

kgrgreer committed:
Temporary fix to ViewSpec for menus only.

kgrgreer committed:
Temporary fix for ServletHandler breaking everything.

mdittmer committed:
Change RestDAOHandler port

kgrgreer committed:
Change Test failed/passed to Longs to work around server class cast bug.

kgrgreer committed:
Merge branch 'master' of https://github.com/foam-framework/foam2

kgrgreer committed:
Add runFailedTests action to TestBorder.

kgrgreer committed:
Add temporary fix to BrowserView until ViewSpec is fixed.

mdittmer committed:
Drop debugging code from pipeline test

mdittmer committed:
Fix XSS vulnerability of foam.net.node.Handler (#572)

Add support for HTML-escaping, and use it in error message handlers

adamvy committed:
Initial work on exception support in Stubs/Skeletons.

mdittmer committed:
SyncDAO/VersionNoDAO to use generated "Versioned" classes

This change generates "versioned" classes with properties `version_` and
`deleted_` for interal use with `SyncDAO` and `VersionNoDAO`. Currently,
`VersionNoDAO` expects a versioned `of` (and does not automatically create
one, but maybe it should?).

The value of this change is that client code outside of the `SyncDAO` can
operate on classes that do not contain versioning information; this becomes
an implementation detail hidden inside:
(Sync DAO ==> ... VersionNoDAO => ... ==> <storage mechanism>).

mdittmer committed:
Drop debugging code

mdittmer commented:
@jacksonic PTAL. I am not as happy with this as I thought I would be for two reasons:

1. Look at the stress test code. It feels like enough layers need to know about versioning that it is hardly worth keeping it out of the one layer that doesn't. Also, the developer now needs to keep track of which layers do and don't know about versioning. Maybe I just feel that way because what's being tested is the interface between versioned and unversioned objects; WDYT?
2. I have a use case that uses two layers of SyncDAO. I'm not sure whether this change is compatible with that use case. (Or at least, `SyncDAO` may need *another* change to detect that `of` is already versioned, and don't bother adapting client objects).

LMK what you think.

kgrgreer committed:
Don't absorb exceptions in ServiceServlet.

kgrgreer committed:
Merge branch 'master' of https://github.com/foam-framework/foam2

kgrgreer committed:
Better exception handling in NanoRouter.

kgrgreer committed:
Start of Temperature view for PMs.

kgrgreer committed:
Temporary fix and formatting.

jacksonic commented on src/foam/dao/SyncDAO.js:
```diff
@@ -112,6 +118,18 @@ foam.CLASS({
+      name: 'versionedOf_',
+      expression: function(of) {
+        return this.versionedClassFactory_.get(of);
+      }
```
I think what's making this seem complicated is the inclusion of the versioning adapter here. If a DAO decorator was adapting the instances on the client end of SyncDAO, then everything in here could be implemented as before, and a facade or factory class could configure the parts of the sync system to use the versioned class. That would mean the versionedClassFactory would reside in the facade, using a generic TypeAdapterDAO + generic (but expecting intrinsic versioning) sync system

mdittmer replied:
I agree. At the moment, I'm using this `SyncDAO` implementation (so I need to land it), but I do not depend on `SyncDAO` owning this adaptation (current usage is `of: theVersionedClass` in server <--> server interactions).

All that is to say, pulling this out to `TypeAdapterDAO` should be a prerequisite to integrating with `EasyDAO`, but I don't have time to implement it right now.

jacksonic replied:
Offline discussion: yes, we'll refactor this stuff out in a future PR

adamvy committed:
Merge branch 'master' of github.com:foam-framework/foam2

mdittmer committed:
VersionedClassFactory: Do not double-version classes

adamvy committed:
Fix syntax error in SkeletonBox.

adamvy committed:
Add RequestResponseClientDAO which simulates events on
put/remove/removeAll.

Switch to RequestResponseClientDAO in nanos Client.js until we add
event support to SkeletonBoxes.

Fixes #568

mdittmer committed:
Nit: Typo in VersionNoDAO error message

mdittmer committed:
Merge branch 'master' into sync-dao__next__code-gen

mdittmer committed:
Merge branch 'datastore-partition-id'

mdittmer committed:
Merge branch 'master' into sync-dao__next__code-gen

mdittmer committed:
Add FUTURE note for pulling versioned/unversioned type adaptation out of SyncDAO

mdittmer commented:
Manually merged to master: https://github.com/foam-framework/foam2/commit/5d61f913e0f9720ebb9f726d8633f47f3781a231
